adam/learner/affordances.py

```diff
@@ -0,0 +1,195 @@
+from abc import ABC
+from typing import AbstractSet, Tuple, Optional, Iterable
+
+from attr import attrs
+from immutablecollections import immutablesetmultidict, immutableset
+
+from adam.learner import (
+    LanguagePerceptionSemanticAlignment,
+    PerceptionSemanticAlignment,
+    SurfaceTemplateBoundToSemanticNodes,
+)
+from adam.learner.attributes import _extract_candidate_attributes
+from adam.learner.perception_graph_template import PerceptionGraphTemplate
+from adam.learner.subset import AbstractTemplateSubsetLearner
+from adam.learner.surface_templates import SLOT1, SurfaceTemplate
+from adam.learner.template_learner import AbstractTemplateLearner
+from adam.perception import MatchMode
+from adam.perception.perception_graph import PerceptionGraph, HAS_PROPERTY_LABEL
+from adam.perception.perception_graph_nodes import CategoricalNode
+from adam.perception.perception_graph_predicates import (
+    AnyObjectPredicate,
+    ObjectSemanticNodePerceptionPredicate,
+)
+from adam.semantics import (
+    SemanticNode,
+    Concept,
+    AffordanceConcept,
+    AffordanceSemanticNode,
+    ActionSemanticNode,
+)
+
+
+class AbstractAffordanceTemplateLearner(AbstractTemplateLearner, ABC):
+    def _can_learn_from(
+        self, language_perception_semantic_alignment: LanguagePerceptionSemanticAlignment
+    ) -> bool:
+        return (
+            len(
+                language_perception_semantic_alignment.perception_semantic_alignment.semantic_nodes
+            )
+            > 1
+        )
+
+    def _preprocess_scene(
+        self, perception_semantic_alignment: PerceptionSemanticAlignment
+    ) -> PerceptionSemanticAlignment:
+        return perception_semantic_alignment
```

Comment:
> Do we we want to change this to filter out all the AffordanceSemanticNode s? If we don't, the affordance learner may learn patterns that refer to its own output, because we enrich with the affordance learner before learning with it.

It's reasonable that we shouldn't be able to learn from the self-affordance node (with the way subset works this shouldn't be possible anyway but I'll see if I can easily add a check for it). It seems reasonable to me that some affordances could be indicated by other affordances but agreed that maybe we just shouldn't include affordance properties in the current iteration. I need to double check if I can make that type of change in the preprocessing or if that will trickle down and remove the semantic nodes from the `perception_semantic_alignment` permanently.

> Separate issue: We probably need to handle temporal scope here to make things work. Namely, we have to learn on dynamic perception graphs because we're learning affordances from observing actions, but I think we want to apply the affordance learner to static perception graphs.

Ooh right, I know the  static -> dynamic is 'solved' because the object learners can handle it but the other direction is less clear. One ugly hack would be to just say that any 'static' scene after the actions learners is just really dynamic so that the edge conversion takes place :) (I don't actually hate this hack because it's still right to say 'ENTIRE_SCENE' temporally for a static instance)

Reply:
> I need to double check

Good idea to check, I'm not sure on this one either.

> ugly hack

Not sure I understand. Is the idea here to convert static graphs into dynamic graphs by wrapping every edge in `ENTIRE_SCENE` scopes? I don't hate this idea either, it seems valid.

Reply:
Correct, just convert a static scene into a dynamic scene (which in effect should wrap every edge with an `ENTIRE_SCENE` temporal scope))

Reply:
For the other learners to incorporate affordances, we would have to always do this conversion, right? Or would we do it only under certain circumstances? I didn't follow that part.

This approach seems fine however I suspect this would require some changes to the object learner, which I suspect makes similar assumptions to the attribute learner.

Reply:
> For the other learners to incorporate affordances, we would have to always do this conversion, right? Or would we do it only under certain circumstances? I didn't follow that part.

We could always make this conversion for the test curriculum, let me thinking about it a little more for the learning curriculum.